### PR TITLE
Introduce unittests for the Value class

### PR DESCRIPTION
**Pull request by japsmits**
# Description

Added a unittest for the Value class as implemented in setup.h and setup.cpp.

I currently have some work on this class, so it's good to have a working unittest first. And since it's my first commit to dosbox-staging, it seems like a reasonable place to start.

In the process, I introduced the Overload pattern, which is added to support.h.

# Manual testing

This is an addition to existing unittests, and it passes as is.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [x] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by johnnovak** [comment]
First of all, welcome to the project @japsmits. I like your enthusiasm 🎉 

> Added a unittest for the Value class as implemented in setup.h and setup.cpp.

That's great, such unit test additions are always welcome.

> I currently have some work on this class, so it's good to have a working unittest first. And since it's my first commit to dosbox-staging, it seems like a reasonable place to start.

I highly suggest checking with us first before doing larger-scale reworks that affect the whole codebase. Configuration handling in one of these areas. We're planning to switch the entire config handling mechanism out to something better—it is still to be decided to what exactly.

Consequently, we're only doing maintenance type of work to the current config stuff as it will go away, eventually.

> In the process, I introduced the Overload pattern, which is added to support.h.

We're not against using some more "advanced" C++ features when it's warranted (I started using `std::variant` and `std::visit` myself a while ago), but in this case the ratio of extra complexity added by these features versus the actual test code is rather high.

Unless I'm missing something, I think you could write the exact same test without visit, variant, templates and the new Overload stuff. It will be a bit more verbose, but simpler overall, so it would be a net win. But feel free to explain what's the bigger picture here, I might be missing it 😄 

As a final comment, we generally try to do mass-scale reformats/non-functional cleanups as separate commits. So in this case, it would be a lot better to split this into two commits: 1) cleanup, 2) adding the new test.

**Comment by japsmits** [comment]
> First of all, welcome to the project @japsmits. I like your enthusiasm 🎉

And thanks for your time and effort in writing a comment on my first ever PR on Github. 😄 

> 
> > Added a unittest for the Value class as implemented in setup.h and setup.cpp.
> 
> That's great, such unit test additions are always welcome.
> 
> > I currently have some work on this class, so it's good to have a working unittest first. And since it's my first commit to dosbox-staging, it seems like a reasonable place to start.
> 
> I highly suggest checking with us first before doing larger-scale reworks that affect the whole codebase. Configuration handling in one of these areas. We're planning to switch the entire config handling mechanism out to something better—it is still to be decided to what exactly.
> 
> Consequently, we're only doing maintenance type of work to the current config stuff as it will go away, eventually.

Agreed. My next commit will be something on the issue-list.

> 
> > In the process, I introduced the Overload pattern, which is added to support.h.
> 
> We're not against using some more "advanced" C++ features when it's warranted (I started using `std::variant` and `std::visit` myself a while ago), but in this case the ratio of extra complexity added by these features versus the actual test code is rather high.
> 
> Unless I'm missing something, I think you could write the exact same test without visit, variant, templates and the new Overload stuff. It will be a bit more verbose, but simpler overall, so it would be a net win. But feel free to explain what's the bigger picture here, I might be missing it 😄 Also, the new implementation will probably just use subclassing for the different types, not chucking every possible type into a Value class which is a bit questionable...

I would propose to keep the Overload struct in place, since it is mentioned in several places when looking for how-to's with respect to std::visit and std::variant. It might come in handy in the future. And I'm still surprised it's not part of std, but that's a different matter.

Would it be convenient to add something to our code documentation about the utilities in support.h? It already contains several convenient and/or useful things. And if so, what would be a good place?

In this case, I had no deeper underlying reason to do it this way other than that it was fun to write. :wink: I'll break up and simplify the testcase into simpler to read pieces, since I see where you're comming from. If no longer needed, it can be removed altogether.

> 
> As a final comment, we generally try to do mass-scale reformats/non-functional cleanups as separate commits. So in this case, it would be a lot better to split this into two commits: 1) cleanup, 2) adding the new test.

I realised a bit too late where this was comming from, but I'll split up the commit.

Actual processing of this comment will come next week, since we have a birthday-girl this weekend. 🎂

File: include/support.h [hunk 1]
```diff
@@ -330,4 +334,23 @@ template <typename T>
 std::pair<std::unique_ptr<T[]>, T*> make_unique_aligned_array(
         const size_t byte_alignment, const size_t req_elems, const T& init_val = {});
 
+// Allow for overload resolition for use with std::visit in combination with an
```
Comment (japsmits) [inline include/support.h:337]:
```suggestion
// Allow for overload resolution for use with std::visit in combination with an
```

**Comment by FeralChild64** [comment]
Actually, I have a plan to introduce the _overload_ pattern too - for my upcoming `dir` command rework:

https://github.com/dosbox-staging/dosbox-staging/blob/2a3b319e5aa9ac58139339529a0e1c25e5fee4d3/src/shell/shell_cmds.cpp#L53

**Comment by johnnovak** [comment]
> Actually, I have a plan to introduce the _overload_ pattern too - for my upcoming `dir` command rework:
> 
> https://github.com/dosbox-staging/dosbox-staging/blob/2a3b319e5aa9ac58139339529a0e1c25e5fee4d3/src/shell/shell_cmds.cpp#L53

Yeah, that's a more justified use case 😄

File: include/support.h [hunk 2]
```diff
@@ -330,4 +330,23 @@ template <typename T>
...
+//
+// std::variant<bool, int> data = true;
+// std::string typename = std::visit(Overload{
+//     [](bool) { return "bool"; },
+//     [](int)  { return "int"; },
+// }, data);
+//
+// typename now contains the value "bool".
+template <typename... Ts>
+struct Overload : Ts... {
```
Comment (johnnovak) [inline include/support.h:350]: minor: Probably lowercase `overload` would be better?

File: include/support.h [hunk 3]
```diff
@@ -330,4 +330,23 @@ template <typename T>
 std::pair<std::unique_ptr<T[]>, T*> make_unique_aligned_array(
         const size_t byte_alignment, const size_t req_elems, const T& init_val = {});
 
+// Allow for overload resolition for use with std::visit in combination with an
+// std::variant. A list of functors for each type must be specified.
```
Comment (johnnovak) [inline include/support.h:338]: Maybe include the comment from the commit message here? Not everybody knows what a "functor" is, I think, so this is a bit cryptic.

File: include/support.h [hunk 4]
```diff
@@ -330,4 +330,23 @@ template <typename T>
...
+//
+// Example:
+//
+// std::variant<bool, int> data = true;
+// std::string typename = std::visit(Overload{
+//     [](bool) { return "bool"; },
+//     [](int)  { return "int"; },
+// }, data);
+//
+// typename now contains the value "bool".
```
Comment (johnnovak) [inline include/support.h:348]: Nice example 🎉

**Review by johnnovak — COMMENTED** [review]
Looks good now @japsmits, thanks for making these changes.

Yes, the Overload stuff can stay, apparently @FeralChild64 is planning to put it to good use later.

I've made one small remark about the comments, and the commit messages need a bit of a cleanup. I recommend reading the [How to Write a Git Commit Message](https://cbea.ms/git-commit/) article, we follow those recommendations. But here's in a nutshell what to do:

Your comment commit messages:

```
Introduced the Overload struct
Automatic reformatting
Introduce unittests for the Value class
Automatic reformatting
Typo was fixed
```

After normalising the grammar a bit and squashing the "Typo was fixed" into the first commit, the final result should look something like this:

```
Introduce the Overload struct
Reformat
Introduce unit tests for the Value class
Reformat
```

Thanks @japsmits, we're almost there, this is all just minor stuff 😄

**Comment by japsmits** [comment]
@johnnovak : Thanks again! :smile:

After convincing Git of what I wanted, it looks good now. Could you do one final check? (And for the sake of procedure: who resolves conversations: the initiating author or the one who addresses the point made?)

**Comment by johnnovak** [comment]
> @johnnovak : Thanks again! 😄
> 
> After convincing Git of what I wanted, it looks good now. Could you do one final check? (And for the sake of procedure: who resolves conversations: the initiating author or the one who addresses the point made?)

Thanks @japsmits. In your first few contributions it's probably best to let the reviewer resolve the comments, but we don't have hard rules about it 😄

**Comment by johnnovak** [comment]
Don't worry about the MSVC situation, the CI situation is in a bit of a flux right now.

We'll merge this once the 0.81 final is out, so in a few week's time (we're in a code freeze period; only changes that contribute to the final should go into `main`, generally).

**Comment by japsmits** [comment]
> Don't worry about the MSVC situation, the CI situation is in a bit of a flux right now.
> 
> We'll merge this once the 0.81 final is out, so in a few week's time (we're in a code freeze period; only changes that contribute to the final should go into `main`, generally).

Okay. :smile:. Feel free to merge it whenever it suits.

**Comment by johnnovak** [comment]
> > Don't worry about the MSVC situation, the CI situation is in a bit of a flux right now.
> > We'll merge this once the 0.81 final is out, so in a few week's time (we're in a code freeze period; only changes that contribute to the final should go into `main`, generally).
> 
> Okay. 😄. Feel free to merge it whenever it suits.

We'll need to ask you to rebase when the time comes. We'll ping you! 😄

**Comment by johnnovak** [comment]
@japsmits Can you rebase to main, please? Then it can be merged.

**Comment by japsmits** [comment]
@johnnovak Done! 😄

File: include/support.h [hunk 5]
```diff
@@ -330,4 +330,23 @@ template <typename T>
...
+template <typename... Ts>
+struct Overload : Ts... {
+	using Ts::operator()...;
+};
+template <class... Ts>
+Overload(Ts...) -> Overload<Ts...>;
+
```
Comment (weirddan455) [inline include/support.h:355]: Ok, so you wrote this templated thing but are you actually using it anywhere?  I'm a little confused about this part.

**Comment by weirddan455** [comment]
I had to hit the button so CI will run since it is @japsmits first PR to the project.  Looks like @johnnovak already reviewed it so if he's happy, I think this can be merged.

I was just a little confused because I didn't see a use case for that template but more unit tests are always good.  Thanks for your first contribution :+1:

File: include/support.h (hunk 5 above)
Comment (johnnovak) [reply include/support.h:355]: We agreed it can go in. @FeralChild64 will use it later, and it's handy to have it.

**Comment by johnnovak** [comment]
Thanks for the extra review @weirddan455 , appreciated.

**Pull request merged**